examples/cim/gnn/actor_critic.py

```diff
@@ -0,0 +1,190 @@
+                                        self._device)
+
+        # train actor network
+        # self._actor_optimizer.zero_grad()
+        # self._critic_optimizer.zero_grad()
+        self._optimizer["a&c"].zero_grad()
+
+        # every port has a value
+        # values.shape: (batch, p_cnt)
+
+        probs, values = self._model_dict["a&c"](obs_batch, a=True, p_idx=p_idx, v_idx=v_idx, c=True)
+        distribution = Categorical(probs)
+        log_prob = distribution.log_prob(action_batch)
+        entropy_loss = distribution.entropy()
+
+        _, values_ = self._model_dict["a&c"](next_obs_batch, c=True)
+        advantage = return_batch + self._value_discount * values_.detach() - values
+
+        if self._entropy_factor != 0:
+            # actor_loss = actor_loss* torch.log(entropy_loss + np.e)
+            advantage[:, p_idx] += self._entropy_factor*entropy_loss.detach()
+
+        actor_loss = - (log_prob*torch.sum(advantage, axis=-1).detach()).mean()
+
+        # actor_loss.backward(retain_graph=True)
+        # self._actor_optimizer["a&c"].step()
+        
+        item_a_loss = actor_loss.item()
+        item_e_loss = entropy_loss.mean().item()
+
+        # train critic network
+        critic_loss = torch.sum(advantage.pow(2), axis=1).mean()
+        # critic_loss.backward()
+        item_c_loss = critic_loss.item()
+        # torch.nn.utils.clip_grad_norm_(self._critic_model.parameters(),0.5)
+        # self._critic_optimizer["a&c"].step()
+        tot_loss = 0.1*actor_loss + critic_loss # - self._entropy_factor * entropy_loss
+        tot_loss.backward()
+        tot_norm = clip_grad.clip_grad_norm_(self._model_dict["a&c"].parameters(), 1)
+        self._optimizer["a&c"].step()
+        return item_a_loss, item_c_loss, item_e_loss, float(tot_norm)
+
+    def set_weights(self, weights):
+        self._model_dict["a&c"].load_state_dict(weights)
+
+    def get_weights(self):
+        return self._model_dict["a&c"].state_dict()
+
+    def _get_save_idx(self, fp_str):
+        return int(fp_str.split('.')[0].split('_')[0])
+
+    def save_model(self, pth, id):
+        if not os.path.exists(pth):
+            os.makedirs(pth)
+        pth = os.path.join(pth, '%d_ac.pkl'%id)
+        torch.save(self._model_dict["a&c"].state_dict(), pth)
+    
+    def _set_gnn_weights(self, weights):
+        for key in weights:
+            if key in self._model_dict["a&c"].state_dict().keys():
+                self._model_dict["a&c"].state_dict()[key].copy_(weights[key])
+
+    def load_model(self, folder_pth, idx=-1):
+        if idx == -1:
+            fps = os.listdir(folder_pth)
+            fps = [f for f in fps if 'ac' in f]
+            fps.sort(key=self._get_save_idx)
+            ac_pth = fps[-1]
+        else:
+            ac_pth = '%d_ac.pkl'%idx
+        pth = os.path.join(folder_pth, ac_pth)
+        with open(pth, 'rb') as fp:
+            weights = torch.load(fp, map_location=self._device)
+        self._set_gnn_weights(weights)
```

Comment:
a new line is needed at the end of the file. https://editorconfig.org/#download can help you automatically add it.